### ISSUE TITLE
Refactor in Formdata_to_model

### ISSUE THREAD

**Issue by rikirenz** [inspirehep/inspire-next#1904]
This issue wants to be a checklist for the refactor in `formdata_to_model`. Most of the fields are quite clear others need explanations. 

- [ ] `extra_comments`:

    - **formdata_to_model**:
        - now: we put it in the`hidden_notes`  (that is not part of the schema)
        - after: it has to be put in the new field `_private_notes`

    - **dojson_rule**:
         - now: not present
         - after: not necessary

- [ ] `title_crossref`:

    - **formdata_to_model**:
        - now: we put it in the `titles` (that is in the schema)
        - after: it should remain here

    - **dojson_rule**:
         - now: not present
         - after: not necessary 

- [ ] `conf_name`:

    - **formdata_to_model**:
        - now: it is in the`hidden_notes`  (that is not part of the schema)
        - after: it has to be put in the new field `_private_notes` 

    - **dojson_rule**:
         - now: not present
         - after: not necessary

- [ ] `title_arXiv`:
    - **formdata_to_model**:
        - now: we put it in the `titles` (that is in the schema)
        - after: it should remain here

    - **dojson_rule**:
         - now: not present
         - after: not necessary

- [ ] `degree_type`:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
         - now: we put this in the `thesis`(that is not part of the schema)
         - after: we should put it in the `thesis_info` field. Am I right? @kaplun 

- [ ] `references`:
    - **formdata_to_model**:
        - now: not present
        - after: not nececssary

    - **dojson_rule**:
         - now: not present
         - after: we should create a rule and put all in the references field 
           (that is part of the schema)

- [ ] `year`:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
         - now: we put it in `publication_info.year`
         - after: it should remain the same

- [ ] `nonpublic_note`:
    - **formdata_to_model**:
        - now: we put it in the`hidden_notes`  (that is not part of the schema)
        - after: it has to be put in the new field `_private_notes`

    - **dojson_rule**:
         - now: not present 
         - after: not necessary

- [ ] `title`:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
        - now: we put it in the `titles` (that is in the schema)
        - after: it should remain here

- [ ] `type_of_doc`:
    - **formdata_to_model**:
        - now: we put it in the `collections` field (that is not in the schema), and only if a thesis
        - after: ? @kaplun 

    - **dojson_rule**:
         - now: not present 
         - after: not necessary

- [ ] `arxiv_id`:
    - **formdata_to_model**:
        - now: we put it in the `external_system_numbers` field (that is not part of the schema)
        - after: we should remove this 

    - **dojson_rule**:
         - now: we put it in `arxiv_eprints` field
         - after: it should remain the same 

- [ ] `note`:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
         - now: we put t in the `public_notes` field (that is part the schema)
         - after: it should remain the same 

- [ ] `experiment`:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
         - now: we put it in the `accelerator_experiments` field (that is part of the schema)
         - after: it should remain the same 

- [ ] `preprint_created`:
    - **formdata_to_model**:
        - now: not present 
        - after: not necessary

    - **dojson_rule**:
         - now: we put it in the `imprints` field (that is part of the schema)
         - after: it should remain the same

- [ ] `conference_id`:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
         - now: we put it in the `publication_info` field (that is part of the schema)
         - after: it should remain the same

- [ ] `issue`:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
         - now: we put it in the `publication_info` field (that is part of the schema)
         - after: it should remain the same

- [ ] `abstract`:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
         - now: we put it in the `abstracts` field (that is part of the schema)
         - after: it should remain the same

- [ ] `thesis_date`:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
         - now: we are put it in the `thesis` field (that is not part of the schema)
         - after: we should put it in the `thesis_info` field. Am I right? @kaplun 

- [ ] `journal_title`:
    - **formdata_to_model**:
        - now: we do a normalization in order to get the shortest name for a given journal
        - after: it should remain the same

    - **dojson_rule**:
         - now: we put it in the `publication_info` field (that is part of the schema)
         - after: it should remain the same

- [ ] `collaboration`:
    - I think that for this field we need a clarification @kaplun  @michamos 

- [ ] `volume`:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
         - now: we put it in the `publication_info` field (that is part of the schema)
         - after: it should remain the same

- [ ] `subject_term`:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
         - now: we put it in the `inspire_categories` field (that is part of the schema)
         - after: it should remain the same

- [ ] `supervisors`:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
         - now: we put it in the `thesis_supervisors` field (that is not part of the schema)
         - after: we should put it in the `authors` field.  Am I right?  @kaplun  

- [ ] `report_numbers`:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
         - now: we put it in the `report_numbers` field (that is part of the schema)
         - after: it should remain the same

- [ ] `authors`:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
        - now: we put it in the `authors` field (that is part of the schema)
        - after: it should remain the same

- [ ] defense_date:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
        - now: we put it in the `thesis` field (that is not part of the schema)
        - after: we should put it in the `thesis_info` field.  Am I right?  @kaplun 

- [ ] institution:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
        - now: we put it in the `thesis` field (that is not part of the schema)
        - after: we should put it in the `thesis_info` field.  Am I right?  @kaplun 

- [ ] categories:
    - **formdata_to_model**:
         - now: we use `arxiv_eprints.categories` it to do evaluations on other fields 
         - after: it should remain the same

    - **dojson_rule**:
        - now: we put it in the `inspire_categories` field (that is part of the schema)
        - after: it should remain the same
        - now: we put it in the `arxiv_eprints.categories` field (that is part of the schema)
        - after: it does not make sense because we have an enum that defines this. need explanation ?   @kaplun 

- [ ] doi:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
        - now: we put it in the `dois` field (that is part of the schema)
        - after: it should remain the same

- [ ] language:
    - **formdata_to_model**:
        - now: we do evaluations if  the language is `oth`
        - after: it should remain the same

    - **dojson_rule**:
        - now: we put it in the `languages` field (that is part of the schema)
        - after: it should remain the same

- [ ] url:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
        - now: we put it in the `url` field (that is part of the schema)
        - after: it should remain the same

- [ ] page_range_article_id:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
        - now: we put it in the `publication_info` field (that is part of the schema)
        - after: it should remain the same

- [ ] title_translation:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
        - now: we put it in the `title` field (that is part of the schema)
        - after: it should remain the same

- [ ] additional_url:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
        - now: we put it in the `url` field (that is part of the schema)
        - after: it should remain the same

- [ ] license_url:
    - **formdata_to_model**:
        - now: not present
        - after: not necessary

    - **dojson_rule**:
        - now: we put it in the `license.url` field (that is part of the schema)
        - after: it should remain the same

- [ ] other_language:
    - **formdata_to_model**:
        - now: we put it in the `languges` field (that is part of the schema)
        - after: it should remain the same

    - **dojson_rule**:
        - now: not present
        - after: not necessary

**Comment by jacquerie** [comment]
Wait. It's the other way around: most of the stuff should actually be done as DoJSON rules, not in `formdata_to_model`, which should disappear (or become post-DoJSON filters when they are necessary).

**Comment by rikirenz** [comment]
I would like to fix first the schema and validate the JSON that comes from the form. And after move all checks that `formdata_to_model` is doing in filters.

if there is something that I can do in `dojson` rather than in `formdata_to_model` I will move it. But first I would like to have the test and a valid JSON

**Comment by kaplun** [comment]
@jacquerie today at standup we discussed about inspirehep/inspire-next#1904
There it appears that dojsonis not suitable for many-to-one field pattern. On the other hand this seems like the usecase of record builders that we have discussed many times but always postponed due to complexity.
However, since @rikirenz is refactoring formdata_to_model he could actually just use a builder that he would create for that purpose.
@jmartinm has an example of builder (for reference) that @rikirenz could use as blueprint to implement such builder.
Then the builder could be re-used in hewcrawl.
The only question is: how much this would delay delivering the milestone?

**Comment by rikirenz** [comment]
actually I do not know. I will try to do this as fast as I can.

**Comment by jacquerie** [comment]
> There it appears that dojsonis not suitable for many-to-one field pattern.

I don't see many examples of this happening in https://github.com/inspirehep/inspire-next/blob/3383a0a9977001abf7fef480fc559fbb9a786f4d/tests/unit/literaturesuggest/test_literaturesuggest_tasks.py. When that happens (for example, for the `publication_info` stuff), you can add a post-DoJSON filter that makes a decision based on various fields.

**Comment by kaplun** [comment]
How do you see from the tests? With the current implementation this appear easily in all the `dojson`s rules that have to use `self`:
* https://github.com/inspirehep/inspire-next/blob/master/inspirehep/modules/literaturesuggest/dojson/fields/literature.py#L57
* https://github.com/inspirehep/inspire-next/blob/master/inspirehep/modules/literaturesuggest/dojson/fields/literature.py#L96
* https://github.com/inspirehep/inspire-next/blob/master/inspirehep/modules/literaturesuggest/dojson/fields/literature.py#L122
* https://github.com/inspirehep/inspire-next/blob/master/inspirehep/modules/literaturesuggest/dojson/fields/literature.py#L131
* https://github.com/inspirehep/inspire-next/blob/master/inspirehep/modules/literaturesuggest/dojson/fields/literature.py#L141
* https://github.com/inspirehep/inspire-next/blob/master/inspirehep/modules/literaturesuggest/dojson/fields/literature.py#L153
* ...

**Comment by jacquerie** [comment]
You see that from the tests, because the test detail what goes in, and what goes out. For example, [this test](https://github.com/inspirehep/inspire-next/blob/2a1f24d62f5e35b938574a06a52ebf24275712ff/tests/unit/literaturesuggest/test_literaturesuggest_tasks.py#L405-L426) proves that the key `languages` in the record data model is generated from the key `language` in the form data model (`type_of_doc` and `title` are there because `formdata_to_model` implicitly requires them, but they are not needed in the computation).

**Comment by jacquerie** [comment]
> How do you see from the tests? With the current implementation this appear easily in all the `dojson`s rules that have to use `self`:

This heuristic is completely wrong.

> * https://github.com/inspirehep/inspire-next/blob/master/inspirehep/modules/literaturesuggest/dojson/fields/literature.py#L57

This rule is generating `arxiv_eprints` from `arxiv_id`.

> * https://github.com/inspirehep/inspire-next/blob/master/inspirehep/modules/literaturesuggest/dojson/fields/literature.py#L96

This rule is generating `inspire_categories` from `categories`. Then, `arxiv_eprints` are incorrectly modified, but this mirrors the same thing that happens in the main HEP DoJSON model: you can decide the Inspire category only at the end.

> * https://github.com/inspirehep/inspire-next/blob/master/inspirehep/modules/literaturesuggest/dojson/fields/literature.py#L122

This is part of the `publication_info` stuff that I mentioned above.

> * https://github.com/inspirehep/inspire-next/blob/master/inspirehep/modules/literaturesuggest/dojson/fields/literature.py#L131

This rule is generating `imprints` from `preprint_created`.

> * https://github.com/inspirehep/inspire-next/blob/master/inspirehep/modules/literaturesuggest/dojson/fields/literature.py#L141
> * https://github.com/inspirehep/inspire-next/blob/master/inspirehep/modules/literaturesuggest/dojson/fields/literature.py#L153

Finally! These rules are generating `thesis` from `defense_date`, `degree_type`, `institution`. But, because `thesis` is an object and they modify independent keys, they don't conflict with each other. So these are also wrong.

**Comment by kaplun** [comment]
> This is part of the publication_info stuff that I mentioned above.
> This rule is generating imprints from preprint_created.
> These rules are generating thesis from defense_date, degree_type, institution.

Correct. As you see `dojson` imposes quite some complexity (even just considering the amount of lines of code involved) even when the output is a simple object and not a list of objects. Instead for each one of the above cases, thinking with the builder pattern you would have only one method per outgoing object that can take at the same time many input keys. This can really reduce the codebase.

**Comment by jacquerie** [comment]
Given that the real discussion behind this issue happened elsewhere, this issue can be closed because it is no longer useful (you decided to go with another technology).
